scipy/integrate/_quadrature.py

```diff
@@ -822,6 +825,305 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
     return result
 
 
+def _cumulatively_sum_simpson_subintegrals(
+    sub_integrals_h1: np.ndarray,
+    sub_integrals_h2: np.ndarray,
+) -> np.ndarray:
+    """Calculate cumulative sum of Simpson subintegrals.
+
+    Takes as input the estimated Simpson integrals over h1 and
+    h2 subintervals. Returns the cumulative sum using
+    composite Simpson's rule. Assumes the axis of summation is -1.
+    """
+    # Set integral of odd h1 intervals = 0
+    sub_integrals_h1[..., 1::2] = 0.0
+    # Set integral of odd h2 intervals = 0, except for the last interval
+    sub_integrals_h2[..., 1:-1:2] = 0.0
+    sub_integrals = np.concatenate(
+        [
+            # Integral over first subinterval can only be calculated from 
+            # formula for h1
+            sub_integrals_h1[..., 0:1],
+            # Combine sub_integral arrays (at every index, i, either
+            # sub_integrals_h1[i] = 0 or sub_integrals_h2[i] = 0)
+            (sub_integrals_h1[..., 1:] + sub_integrals_h2[..., :-1]),
+            # Integral over last subinterval can only be calculated from 
+            # formula for h2
+            sub_integrals_h2[..., -1:],
+        ],
+        axis=-1,
+    )
```

Comment:
Isn't this all just like:
```python3
sub_integrals = np.append(sub_integrals_h1, sub_integrals_h2[-1:], axis=-1)
sub_integrals[1::2] = sub_integrals_h2[::2]
```
?

Also, it looks like we're throwing away half the values that were calculated. Why don't we just calculate the ones that are needed?

Reply:
Optimized this further. I used np.concatenate instead of np.append as I found it was more efficient.

> it looks like we're throwing away half the values that were calculated. Why don't we just calculate the ones that are needed?

I have refactored the internal functions to only integrate the intervals required. 
It is slightly more efficient than if we just calculate the entire array and discard half (see optimized version 1 vs version 2 at the end of [this gist](https://gist.github.com/nprav/1c8a82738cb4f8ceffe740a5500b8d61)). 
But, it does make the internal functions and function signatures a bit more complicated.
Let me know if you can think of a better way to implement this.

Reply:
Let's do optimized version 1. Simplicity is what's most important now, if we're trying to get this merged before branching 1.13.

(I considered starting with an array of zero of the right shape then populating it, as you have. I ended up with what I wrote because it had the same complexity and was simpler, which was the motivation. But what you have in optimized verison 1 is fine. You can use [`np.empty`](https://numpy.org/doc/stable/reference/generated/numpy.empty.html) instead of `np.zeros`, even though it rarely make a practical difference.)